pyensembl/genome.py

```diff
@@ -0,0 +1,848 @@
+
+        contig : str, optional
+            Restrict query to particular contig
+
+        strand : str, optional
+            Restrict results to "+" or "-" strands
+
+        Returns a list constructed from query results.
+        """
+        # since we're constructing a list, rather than a DataFrame,
+        # we're going to store it using pickling rather than the Pandas
+        # CSV serializer. Change the default extension from ".csv" to ".pickle"
+        pickle_path = self.gtf.local_data_file_path(
+            feature=feature,
+            column=column,
+            contig=contig,
+            strand=strand,
+            distinct=distinct,
+            extension=".pickle")
+
+        def run_query():
+            results = self.db.query_feature_values(
+                column=column,
+                feature=feature,
+                distinct=distinct,
+                contig=contig,
+                strand=strand)
+            assert isinstance(results, list), \
+                "Expected list from Database.query_feature_values, got %s" % (
+                    type(results))
+            return results
+
+        return cached_object(pickle_path, compute_fn=run_query)
+
+    def install(self):
+        """
+        Explicitely download and index any data for this genome source that
+        is not yet downloaded and/or indexed.
+        """
+        self.download(force=False)
+        self.index(force=False)
+
+    def index(self, force=True):
+        """
+        Create databases and indices for all data for this genome source
+
+        Parameters
+        ----------
+        force : bool
+            Recreate databases even if already indexed (default = True)
+
+        Raises an error if data is not downloaded.
+        """
+        self.db.create(force=force)
+        if self.transcript_sequences:
+            self.transcript_sequences.index(force=force)
+        if self.protein_sequences:
+            self.protein_sequences.index(force=force)
+
+    def transcript_sequence(self, transcript_id):
+        """Return cDNA nucleotide sequence of transcript, or None if
+        transcript doesn't have cDNA sequence.
+        """
+        assert self.transcript_sequences, (
+            "This genome source does not include transcript FASTA data: %s"
+            % self.genome_source)
+        if self.only_human:
+            require_human_transcript_id(transcript_id)
+        return self.transcript_sequences.get(transcript_id)
+
+    def protein_sequence(self, protein_id):
+        """Return cDNA nucleotide sequence of transcript, or None if
+        transcript doesn't have cDNA sequence.
+        """
+        assert self.protein_sequences, (
+            "This genome source does not include protein FASTA data: %s"
+            % self.genome_source)
+        if self.only_human:
+            require_human_protein_id(protein_id)
+        return self.protein_sequences.get(protein_id)
+
+    def download(self, force=True):
```

Comment:
Discussed offline: `GenomeSource` should either do more or not exist, though we'll keep it for this PR.

It does seem nice to separate `index`/`install`/`download` from all the gene related methods, so in that sense moving them to `GenomeSource` would be cool.

That said, if someone just wants to run `EnsemblRelease(77).install()`, it would be annoying if they actually needed to touch a different class (`EnsemblReleaseSource`) as well.